qiskit/qasm2/__init__.py

```diff
@@ -0,0 +1,523 @@
+===========
+
+This module contains two public functions, both of which create a :class:`.QuantumCircuit` from an
+OpenQASM 2 program. :func:`load` takes a filename, while :func:`loads` takes the program itself as a
+string.  Their internals are very similar, so both offer almost the same API.
+
+.. autofunction:: load
+
+.. autofunction:: loads
+
+Both of these loading functions also take an argument ``include_path``, which is an iterable of
+directory names to use when searching for files in ``include`` statements.  The directories are
+tried from index 0 onwards, and the first match is used.  The import ``qelib1.inc`` is treated
+specially; it is always found before looking in the include path, and contains exactly the content
+of the `paper describing the OpenQASM 2 language <https://arxiv.org/abs/1707.03429>`__.  The gates
+in this include file are mapped to standard gates provided by Qiskit.
+
+.. _qasm2-custom-instructions:
+
+Specifying custom instructions
+------------------------------
+
+You can extend the quantum components of the OpenQASM 2 language by passing an iterable of
+information on custom instructions as the argument ``custom_instructions``.  In files that have
+compatible definitions for these instructions, the given ``constructor`` will be used in place of
+whatever other handling :mod:`qiskit.qasm2` would have done.  These instructions may optionally be
+marked as ``builtin``, which causes them to not require an ``opaque`` or ``gate`` declaration, but
+they will silently ignore a compatible declaration.  Either way, it is an error to provide a custom
+instruction that has a different number of parameters or qubits as a defined instruction in a parsed
+program.  Each element of the argument iterable should be a particular data class:
+
+.. autoclass:: CustomInstruction
+
+.. _qasm2-custom-classical:
+
+Specifying custom classical functions
+-------------------------------------
+
+Similar to the quantum extensions above, you can also extend the processing done to classical
+expressions (arguments to gates) by passing an iterable to the argument ``custom_classical`` to either
+loader.  This needs the ``name`` (a valid OpenQASM 2 identifier), the number ``num_params`` of
+parameters it takes, and a Python callable that implements the function.  The Python callable must
+be able to accept ``num_params`` positional floating-point arguments, and must return a float or
+integer (which will be converted to a float).  Builtin functions cannot be overridden.
+
+.. autoclass:: CustomClassical
+
+.. _qasm2-strict-mode:
+
+Strict mode
+-----------
+
+Both of the loader functions have an optional "strict" mode.  By default, this parser is a little
+bit more relaxed than the official specification: it allows trailing commas in parameter lists;
+unnecessary (empty-statement) semicolons; the ``OPENQASM 2.0;`` version statement to be omitted; and
+a couple of other quality-of-life improvements without emitting any errors.  You can use the
```

Comment:
```suggestion
a couple of other flexibilities without emitting any errors.  You can use the
```

Reply:
I left this as-is because I didn't think it was too meaningful a change.